File: Leetcode/1260. Shift 2D Grid/solution.rs
impl Solution {
    pub fn shift_grid(mut grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let rows = grid.len();
        let cols = grid[0].len();
        let total = rows * cols;
        let k = (k as usize) % total;

        let mut start = 0;
        let mut count_visited = 0;

        while count_visited < total {
            let mut current_index = start;
            let mut current_value = grid[current_index / cols][current_index % cols];

            loop {
                let next_index = (current_index + k) % total;
                let next_value = grid[next_index / cols][next_index % cols];

                grid[next_index / cols][next_index % cols] = current_value;
                current_index = next_index;
                current_value = next_value;
                count_visited += 1;

                if current_index == start {
                    break;
                }
            }

            start += 1;
        }

        grid
    }
}

File: Leetcode/1260. Shift 2D Grid/solution.py
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        if not grid or not grid[0]:
            return grid
        
        rows = len(grid)
        cols = len(grid[0])
        total = rows * cols
        k = k % total  # Only shift necessary positions
        
        start = 0
        count_visited = 0
        
        while count_visited < total:
            current_index = start
            current_value = grid[current_index // cols][current_index % cols]
            
            while True:
                # Calculate new position
                next_index = (current_index + k) % total
                next_value = grid[next_index // cols][next_index % cols]
                
                # Place the current value into its new position
                grid[next_index // cols][next_index % cols] = current_value
                current_index = next_index
                current_value = next_value
                count_visited += 1
                
                # If we are back to the start, break the loop
                if current_index == start:
                    break
            
            start += 1  # Move to the next starting position to check other cycles
        
        return grid
